File: odt-search/src/main/java/pl/edu/agh/odt/logic/FileFinder.java
package pl.edu.agh.odt.logic;

import pl.edu.agh.odt.result.FileSearchResult;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class FileFinder {

    private final List<String> exceptions;
    private final String rootPath;
    private final List<String> phrases;

    public FileFinder(String rootPath, List<String> phrases){
        this.exceptions = new ArrayList<>();
        this.rootPath = rootPath;
        this.phrases = phrases;
    }

    public List<String> getExceptions(){
        return exceptions;
    }

    public List<FileSearchResult> findInDirectory() throws IOException {
        return getOdtPathStream().map(filePath -> {
            try {
                return OdtSearcher.phraseOccurrences(phrases, filePath);
            } catch (Exception e) {
                exceptions.add(e.getMessage());
                return null;
            }
        })
        .filter(Objects::nonNull)
        .collect(Collectors.toList());
    }

    private Stream<String> getOdtPathStream() throws IOException {
       return Files.walk(Path.of(rootPath))
        .filter(file -> file.getFileName().toString().endsWith(".odt") || file.getFileName().toString().endsWith(".odf"))
        .map(Path::toString);
    }



}

File: ocr-search/src/main/kotlin/pl/edu/agh/ocr/logic/FileFinder.kt
package pl.edu.agh.ocr.logic

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import pl.edu.agh.ocr.data.OcrResult
import java.nio.file.Files
import java.nio.file.Path
import java.util.stream.Collectors
import java.util.stream.Stream

@Component
class FileFinder(@Autowired val logger: Logger, private val ocrWorker: OcrWorker) {

    private val extensions: List<String> = listOf(".png", ".jpg", ".jpeg")

    fun extractFromDirectory(rootPath: String): Pair<List<OcrResult>, List<String>> {
        logger.info("Starting extracting text from images in path: '$rootPath'")
        val ocrResults = mutableListOf<OcrResult>()
        val exceptions = mutableListOf<String>()
        val stream = getOcrPathStream(rootPath)
        val mapped = stream.map(ocrWorker::getExtractionResult)
        mapped.forEach { pair ->
            run {
                ocrResults.addAll(pair.first)
                exceptions.addAll(pair.second)
            }
        }
        logger.info("Finished extracting text from images in path: '$rootPath'")
        return Pair(ocrResults, exceptions)
    }

    private fun getOcrPathStream(rootPath: String): List<String> = Files.walk(Path.of(rootPath))
        .filter(this::isOcrFile)
        .map(Path::toString)
        .collect(Collectors.toList())

    private fun isOcrFile(filePath: Path): Boolean {
        val fileName = filePath.fileName.toString()
        return extensions.any { ex -> fileName.endsWith(ex) }
    }
}